### PR TITLE
TEL review of day 4 homework

### PR DESCRIPTION
**Pull request by tomlatham**
Very good attempt, it all worked exactly as expected.  A couple of small improvements here and there that you can find in the line-by-line comments.  There were also no tests, so I've added a very basic one.  One would need a few to properly test all the edge cases.  Let me know if you have any questions.

### CODE REVIEW

File: MPAGSCipher/PlayfairCipher.cpp [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #include "PlayfairCipher.hpp"
 
 PlayfairCipher::PlayfairCipher( const std::string& key )
+{
+	setKey(key);
+}
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: The constructor should just call setKey

File: MPAGSCipher/PlayfairCipher.cpp [hunk 2]
```diff
@@ -1,6 +1,11 @@
...
+
+void PlayfairCipher::setKey( const std::string& key )
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: setKey was not implemented but the constructor implementation was what should have been here, so the edit it straightforward!

File: MPAGSCipher/PlayfairCipher.cpp [hunk 3]
```diff
@@ -45,7 +50,7 @@ PlayfairCipher::PlayfairCipher( const std::string& key )
 		bool isDupli = std::find(soFar.begin(),
 				soFar.end(),
 				c) != soFar.end();
-		soFar += c;
+		if ( !isDupli ) { soFar += c; }
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Only add if not before encountered to slightly improve efficiency.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 4]
```diff
@@ -57,16 +62,12 @@ PlayfairCipher::PlayfairCipher( const std::string& key )
 	key_.erase(iter2,key_.end()); //Erase the duplicates
 
 	//store the coords of each letter
-	using mapEntry = std::pair< char,std::vector<size_t> >;	
-	using mapEntryCo = std::pair < std::vector<size_t>,char >;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: These typedefs aren't needed due to below modifications.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 5]
```diff
@@ -57,16 +62,12 @@ PlayfairCipher::PlayfairCipher( const std::string& key )
...
-
+	alphaGrid_.clear();
+	coordGrid_.clear();
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Need to clear the maps before populating (does nothing the first time this function is called but is needed for any subsequent calls).

File: MPAGSCipher/PlayfairCipher.cpp [hunk 6]
```diff
@@ -57,16 +62,12 @@ PlayfairCipher::PlayfairCipher( const std::string& key )
...
 	for(unsigned long i{0}; i<25; ++i){//can't use i<alphabetLength_ because we're throwing away J :(
-		std::vector<size_t> pos{i/5,i%5};
-		
-		mapEntry Addition{ std::make_pair(key_[i],pos) };
-		mapEntryCo Addition2{ std::make_pair(pos,key_[i]) };
-		alphaGrid_.insert( Addition );
-		coordGrid_.insert( Addition2);
+		PlayfairCoords pos{i/5,i%5};
+		alphaGrid_.insert( std::make_pair(key_[i],pos) );
+		coordGrid_.insert( std::make_pair(pos,key_[i]) );
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Can simplify this a bit. Here I'm using a type definition that I've placed in the class declaration - see the header file changes below.  Can also put the make_pair directly in the insert call - it's still perfectly readable, at least in my opinion.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 7]
```diff
@@ -77,7 +78,7 @@ PlayfairCipher::PlayfairCipher( const std::string& key )
 	for(auto p : alphaGrid_)
 	{
 		std::cout << "Letter:" << p.first
-			  << "|Pos:(" << p.second[0] << "," << p.second[1]  << ")" << std::endl;
+			  << "|Pos:(" << p.second.first << "," << p.second.second  << ")" << std::endl;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Have changed the type holding the coordinates from vector<size_t> to pair<size_t,size_t> since there will only ever be two entries.  So, [0] becomes .first and [1] becomes .second here and in several other places.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 8]
```diff
@@ -141,71 +142,55 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText,
 		return outputText;
 	}
 
+	const size_t direction { cipherMode == CipherMode::Encrypt ? 1u : 4u };
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: This shift variable can be defined outside the loop since it never changes.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 9]
```diff
@@ -141,71 +142,55 @@ std::string PlayfairCipher::applyCipher( const std::string& inputText,
...
-
-			outputText[i+1] = coordGrid_.find(transformedCoords)->second;
+			transformedCoords2.first  = l2_iter->second.first;
+			transformedCoords2.second = l1_iter->second.second;
 		}
-	}
 
-	
+		outputText[i]   = coordGrid_.find(transformedCoords1)->second;
+		outputText[i+1] = coordGrid_.find(transformedCoords2)->second;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Moved these lines here (by adding a variable to hold the second set of transformed coords) since they were the same in the if, else if, and else cases.

File: MPAGSCipher/PlayfairCipher.hpp [hunk 10]
```diff
@@ -17,7 +17,7 @@
 
 /**
  * \class PlayfairCipher
- * \breif Encrypt or decrypt text using the Playfair cipher with a given key
+ * \brief Encrypt or decrypt text using the Playfair cipher with a given key
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Typo

File: MPAGSCipher/PlayfairCipher.hpp [hunk 11]
```diff
@@ -35,17 +35,18 @@ class PlayfairCipher{
 		/// If no key is given, a null cipher is used (no ciphering)
 		bool nullCipher_{false};
 
+		/// Type definition for the coordinates in the 5x5 table
+		using PlayfairCoords = std::pair<size_t,size_t>;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Here's the type definition of the coordinates.
By placing it here, within the private part of the class declaration, it can be used everywhere within the implementation of the class (but not outside, since it is an implementation detail).

File: MPAGSCipher/PlayfairCipher.hpp [hunk 12]
```diff
@@ -35,17 +35,18 @@ class PlayfairCipher{
...
+
 		/// 5x5 grid to be used for playfair cihper in both directions
-		std::map< char,std::vector<size_t> > alphaGrid_;
-		std::map< std::vector<size_t>,char > coordGrid_;
+		std::map< char,PlayfairCoords > alphaGrid_;
+		std::map< PlayfairCoords,char > coordGrid_;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Can use the type def here too.
